File: userdata/userdata.go
package userdata

import (
	"EnglishProject/cors"
	"EnglishProject/db"
	"EnglishProject/jwt"
	"database/sql"
	"encoding/json"
	"log"
	"net/http"
	"strings"
)

type User struct {
	UserID      int    // 用户ID
	Username    string // 用户ID
	Email       string // 邮箱
	PhoneNumber string // 手机号码
}
type UserData struct {
	ID            int
	UserID        int
	LearningCount int
	Accuracy      float64
	Level         int
	DailyProgress int
	DailyGoal     int
	TotalGoal     int
	TotalProgress int
}
type UserStruct struct {
	UserData UserData `json:"userData"`
	User     User     `json:"userInfo"`
}

func ShowUserDataHandler(w http.ResponseWriter, r *http.Request) {
	cors.EnableCORS(w, r)
	tokenString := r.Header.Get("Authorization")
	if tokenString == "" {
		// 未提供令牌，进行处理或返回未授权的响应
		http.Error(w, "Unauthorized", http.StatusUnauthorized)
		return
	}

	// 去掉 "Bearer " 前缀
	tokenString = strings.TrimPrefix(tokenString, "Bearer ")

	// 调用 ParseJWT 函数来解析和验证 JWT 令牌
	claims, err := jwt.ParseJWT(tokenString)
	if err != nil {
		// 处理令牌无效或解析错误的情况
		http.Error(w, "Token is not valid", http.StatusUnauthorized)
		return
	}

	// 从 claims 中提取用户信息或其他数据
	userIDFloat64 := claims["userID"].(float64)
	userID := int(userIDFloat64)
	log.Printf("userID: %d", userID)
	// 从数据库或其他存储位置检索用户信息
	user, err := getUserInfoFromDatabase(userID)
	if err != nil {
		http.Error(w, "Failed to retrieve user information", http.StatusInternalServerError)
		log.Printf("Error retrieving user information: %v", err)
		return
	}
	userData, err := getUserDataFromDatabase(userID)
	if err != nil {
		http.Error(w, "Failed to retrieve user information", http.StatusInternalServerError)
		log.Printf("Error retrieving userData information: %v", err)
		return
	}
	var userStruct UserStruct
	userStruct.User = *user
	userStruct.UserData = *userData
	// 将用户信息编码为JSON并发送给前端
	responseJSON, err := json.Marshal(userStruct)
	if err != nil {
		http.Error(w, "Failed to serialize JSON response", http.StatusInternalServerError)
		log.Printf("Error serializing JSON: %v", err)
		return
	}
	//log.Printf("json:%s", responseJSON)
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	w.Write(responseJSON)
}
func getUserDataFromDatabase(userID int) (*UserData, error) {
	// 查询数据库以根据 userID 检索用户数据。
	query := "SELECT id, user_id, learning_count, accuracy, level, daily_progress, daily_goal, total_goal, total_progress FROM user_data WHERE user_id = ?"
	row := db.Db.QueryRow(query, userID)

	// 创建 UserData 结构以存储检索到的数据。
	userData := &UserData{}
	err := row.Scan(
		&userData.ID,
		&userData.UserID,
		&userData.LearningCount,
		&userData.Accuracy,
		&userData.Level,
		&userData.DailyProgress,
		&userData.DailyGoal,
		&userData.TotalGoal,
		&userData.TotalProgress,
	)
	if err != nil {
		if err == sql.ErrNoRows {
			log.Printf("找不到 userID 为 %d 的用户数据，将创建新数据", userID)

			// 创建新的用户数据结构
			newUserData := &UserData{
				UserID:        userID,
				LearningCount: 0,
				Accuracy:      0,
				Level:         0,
				DailyProgress: 0,
				DailyGoal:     0,
				TotalGoal:     0,
				TotalProgress: 0,
			}

			// 插入新数据到数据库
			insertQuery := "INSERT INTO user_data (user_id, learning_count, accuracy, level, daily_progress, daily_goal, total_goal, total_progress) VALUES (?, ?, ?, ?, ?, ?, ?, ?)"
			_, err := db.Db.Exec(insertQuery, newUserData.UserID, newUserData.LearningCount, newUserData.Accuracy, newUserData.Level, newUserData.DailyProgress, newUserData.DailyGoal, newUserData.TotalGoal, newUserData.TotalProgress)
			if err != nil {
				return nil, err
			}

			// 返回新创建的用户数据
			return newUserData, nil
		}
		return nil, err
	}

	return userData, nil

}
func ReadPasswordFromRequest(r *http.Request) (string, string, error) {
	var requestData struct {
		OldPassword        string `json:"oldPassword"`
		NewPassword        string `json:"newPassword"`
		ConfirmNewPassword string `json:"confirmNewPassword"`
	}

	decoder := json.NewDecoder(r.Body)
	if err := decoder.Decode(&requestData); err != nil {
		return "", "", err
	}

	return requestData.OldPassword, requestData.NewPassword, nil
}

func getUserInfoFromDatabase(userID int) (*User, error) {
	// 在这里实现从数据库中检索用户信息的逻辑
	query := "SELECT id,username, email, phone FROM users WHERE id = ?"
	var username, email, phone string
	var userid int
	err := db.Db.QueryRow(query, userID).Scan(&userid, &username, &email, &phone)
	if err != nil {
		return nil, err
	}

	// 假设从数据库中获取用户信息
	user := &User{
		UserID:      userid,
		Username:    username,
		Email:       email,
		PhoneNumber: phone,
	}

	return user, nil
}
func ChangePasswordHandler(w http.ResponseWriter, r *http.Request) {
	cors.EnableCORS(w, r)
	tokenString := r.Header.Get("Authorization")
	if tokenString == "" {
		// 未提供令牌，进行处理或返回未授权的响应
		http.Error(w, "Unauthorized", http.StatusUnauthorized)
		return
	}

	// 去掉 "Bearer " 前缀
	tokenString = strings.TrimPrefix(tokenString, "Bearer ")

	// 调用 ParseJWT 函数来解析和验证 JWT 令牌
	claims, err := jwt.ParseJWT(tokenString)
	if err != nil {
		// 处理令牌无效或解析错误的情况
		http.Error(w, "Token is not valid", http.StatusUnauthorized)
		return
	}

	// 从 claims 中提取用户信息或其他数据
	userIDFloat64 := claims["userID"].(float64)
	userID := int(userIDFloat64)
	log.Printf("userID: %d", userID)
	// 获取原密码和新密码
	oldPassword, newPassword, err := ReadPasswordFromRequest(r)
	if err != nil {
		http.Error(w, "Failed to read passwords from the request", http.StatusBadRequest)
		return
	}

	// 查询数据库以获取用户的原密码
	var dbOldPassword string
	qerr := db.Db.QueryRow("SELECT password FROM users WHERE id = ?", userID).Scan(&dbOldPassword)
	if qerr != nil {
		http.Error(w, "Failed to fetch the user's password", http.StatusInternalServerError)
		return
	}

	// 比较提供的原密码和数据库中的原密码
	if oldPassword != dbOldPassword {
		http.Error(w, "Original password does not match", http.StatusUnauthorized)
		log.Printf("输入的 password: %s,数据库的 password: %s", oldPassword, dbOldPassword)
		return
	}

	// 如果原密码匹配，更新数据库中的密码
	_, updateErr := db.Db.Exec("UPDATE users SET password = ? WHERE id = ?", newPassword, userID)
	if updateErr != nil {
		http.Error(w, "Failed to update the password", http.StatusInternalServerError)
		return
	}

	// 密码已成功更改
	// 可以返回成功响应或其他操作
	w.WriteHeader(http.StatusOK)
	w.Write([]byte("Password has been changed successfully"))
}

File: userdata/userdata.js
// 获取用户数据
function getUserData() {
    // 从 localStorage 中获取用户令牌
    var userToken = localStorage.getItem("token");

    // 设置请求头部，包括 Authorization 头部
    var headers = new Headers({
        "Authorization": "Bearer " + userToken,
        "Content-Type": "application/json"
    });

    // 发送请求到后端以获取用户数据
    fetch("http://localhost:8080/showUserData", {
        method: "GET",
        headers: headers
    })
    .then(response => response.json())
    .then(data => {
        // 更新页面上的用户信息
        console.log("user data:", data);
        document.getElementById("userID").textContent = data.userInfo.UserID;
        document.getElementById("userName").textContent = data.userInfo.Username;
        document.getElementById("userEmail").textContent = data.userInfo.Email;
        document.getElementById("userPhone").textContent = data.userInfo.PhoneNumber;
        document.getElementById("userLevel").textContent = data.userData.Level
        document.getElementById("totalProgess").textContent = data.userData.TotalProgess
        document.getElementById("dailyProgess").textContent = data.userData.DailyProgess
        document.getElementById("totalGoal").textContent = data.userData.TotalGoal
        document.getElementById("dailyGoal").textContent = data.userData.DailyGoal
        document.getElementById("accuracy").textContent = data.userData.Accuracy
        document.getElementById("learningCount").textContent = data.userData.LearningCount
    })
    .catch(error => {
        console.error("Error fetching user data:", error);
    });
}

// 添加事件监听器，当页面加载时获取用户数据
window.addEventListener("load", () => {
    getUserData();
});
document.addEventListener("DOMContentLoaded", function() {
    var userToken = localStorage.getItem("token");
    const changePasswordButton = document.getElementById("changePasswordButton");

    changePasswordButton.addEventListener("click", function() {
        // 创建原密码输入框
        const oldPasswordInput = document.createElement("input");
        oldPasswordInput.type = "password";
        oldPasswordInput.placeholder = "原密码";
        oldPasswordInput.id = "oldPasswordInput";
        
        // 创建新密码输入框
        const newPasswordInput = document.createElement("input");
        newPasswordInput.type = "password";
        newPasswordInput.placeholder = "新密码";
        newPasswordInput.id = "newPasswordInput";

        // 创建确认新密码输入框
        const confirmNewPasswordInput = document.createElement("input");
        confirmNewPasswordInput.type = "password";
        confirmNewPasswordInput.placeholder = "确认新密码";
        confirmNewPasswordInput.id = "confirmNewPasswordInput";

        // 创建提交按钮
        const submitButton = document.createElement("button");
        submitButton.textContent = "提交";
        submitButton.id = "submitButton";

        // 添加输入框和提交按钮到用户信息部分
        changePasswordButton.style.display = "none"; // 隐藏"修改密码"按钮
        document.querySelector(".user-info").appendChild(oldPasswordInput);
        document.querySelector(".user-info").appendChild(newPasswordInput);
        document.querySelector(".user-info").appendChild(confirmNewPasswordInput);
        document.querySelector(".user-info").appendChild(submitButton);

        // 监听提交按钮的点击事件
        submitButton.addEventListener("click", function() {
            // 获取原密码、新密码和确认新密码输入框的值
            const oldPassword = document.getElementById("oldPasswordInput").value;
            const newPassword = document.getElementById("newPasswordInput").value;
            const confirmNewPassword = document.getElementById("confirmNewPasswordInput").value;

            // 构建请求数据
            const passwordData = {
                oldPassword: oldPassword,
                newPassword: newPassword,
                confirmNewPassword: confirmNewPassword
            };

            // 发送POST请求到后端，将密码数据放入请求正文
            fetch("http://localhost:8080/changePassword", {
                method: "POST",
                headers: {
                    "Authorization": "Bearer " + userToken,
                    "Content-Type": "application/json"
                },
                body: JSON.stringify(passwordData) // 将密码数据转换为 JSON 字符串
            })
            .then(response => {
                if (response.status === 200) {
                    // 密码修改成功
                    layMsg("密码修改成功", {icon: 1});
                } else {
                    layMsg("密码修改失败", {icon: 2});
                }
            })
            .catch(error => {
                console.error("Error changing password:", error);
            });
        });

        // 移除"修改密码"按钮
        changePasswordButton.style.display = "none";
    });
});
